src/main/scala/edu/uci/eecs/spectralLDA/datamoments/DataCumulantSketch.scala

```diff
@@ -143,45 +182,122 @@ object DataCumulantSketch {
     (eigenVectors(::, 0 until dimK), eigenValues(0 until dimK))
   }
 
-  private def update_firstOrderMoments(dim: Int, Wc: breeze.linalg.DenseVector[Double], len: Double) = {
-    val M1: DenseVector[Double] = Wc / len
-    M1
-  }
+  /** Compute the contribution of the document to the FFT of the sketch of whitened M3
+    *
+    * @param alpha0 Topic concentration
+    * @param W Whitening matrix $W\in\mathsf{R^{V\times k}$, where $V$ is the vocabulary size,
+    *          $k$ is the reduced dimension, $k<V$
+    * @param q Whitened M1, i.e. $W^T M1$
+    * @param n Word count vector for the current document
+    * @param len Total word counts for the current document
+    * @param sketcher The sketching facility
+    * @return The contribution of the document to the FFT of the sketch of whitened M3
+    *         i.e. $E[x_1\otimes x_2\otimes x_3](W^T,W^T,W^T)-
+    *                  \frac{\alpha_0}{\alpha_0+2}\left(E[x_1\otimes x_2\otimes M1]
+    *                                       +E[x_1\otimes M1\otimes x_2]
+    *                                       +E[M1\otimes x_1\otimes x_2]\right)(W^T,W^T,W^T)$
+    *         Refer to Eq (22) in [Wang2015]
+    *
+    * REFERENCES
+    * [Wang2015] Wang Y et al, Fast and Guaranteed Tensor Decomposition via Sketching, 2015,
+    *            http://arxiv.org/abs/1506.04448
+    *
+    */
+  private def update_thirdOrderMoments(alpha0: Double,
+                                       W: DenseMatrix[Double],
+                                       q: DenseVector[Double],
+                                       n: SparseVector[Double],
+                                       len: Double,
+                                       sketcher: TensorSketcher[Double, Double])
+        : DenseMatrix[Complex] = {
+    /* ------------------------------------- */
+
+    // $p=W^T n$, where n is the original word count vector
+    val p: DenseVector[Double] = W.t * n
+
+    // fft of sketch_p, $p=W^T n$, where $n$ is the original word count vector
+    val fft_sketch_p: Seq[DenseMatrix[Complex]] = (0 until 3)
+      .map { sketcher.sketch(p, _) }
```

Comment:
You can [merge those](https://github.com/FurongHuang/SpectralLDA-TensorSpark/pull/11/files#r64030580)